core/search/migrator.py

```diff
@@ -0,0 +1,178 @@
+
+    def __init__(self, fatal: bool, message: str):
+        super().__init__(message)
+        self.fatal = fatal
+
+
+class SearchDocumentReceiverType(ABC):
+    """A receiver of search documents."""
+
+    @abstractmethod
+    def add_documents(
+        self, documents: Iterable[dict]
+    ) -> List[SearchServiceFailedDocument]:
+        """Submit documents to be indexed."""
+
+    @abstractmethod
+    def finish(self) -> None:
+        """Make sure all changes are committed."""
+
+
+class SearchDocumentReceiver(SearchDocumentReceiverType):
+    """A receiver of search documents."""
+
+    def __init__(self, pointer: str, service: SearchService):
+        self._logger = logging.getLogger(SearchDocumentReceiver.__name__)
+        self._pointer = pointer
+        self._service = service
+
+    @property
+    def pointer(self) -> str:
+        """The name of the index that will receive search documents."""
+        return self._pointer
+
+    def add_documents(
+        self, documents: Iterable[dict]
+    ) -> List[SearchServiceFailedDocument]:
+        """Submit documents to be indexed."""
+        return self._service.index_submit_documents(
+            pointer=self._pointer, documents=documents
+        )
+
+    def finish(self) -> None:
+        """Make sure all changes are committed."""
+        self._logger.info("Finishing search documents.")
+        self._service.refresh()
+        self._logger.info("Finished search documents.")
+
+
+class SearchMigrationInProgress(SearchDocumentReceiverType):
+    """A migration in progress. Documents are being submitted, and the migration must be
+    explicitly finished or cancelled to take effect (or not!)."""
+
+    def __init__(
+        self,
+        base_name: str,
+        revision: SearchSchemaRevision,
+        service: SearchService,
+    ):
+        self._logger = logging.getLogger(SearchMigrationInProgress.__name__)
+        self._base_name = base_name
+        self._revision = revision
+        self._service = service
+        self._receiver = SearchDocumentReceiver(
+            pointer=self._revision.name_for_index(base_name), service=self._service
+        )
+
+    def add_documents(
+        self, documents: Iterable[dict]
+    ) -> List[SearchServiceFailedDocument]:
+        """Submit documents to be indexed."""
+        return self._receiver.add_documents(documents)
+
+    def finish(self) -> None:
+        """Finish the migration."""
+        self._logger.info(f"Completing migration to {self._revision.version}")
+        # Make sure all changes are committed.
+        self._receiver.finish()
+        # Create the "indexed" alias.
+        self._service.index_set_populated(self._revision)
+        # Update the write pointer to point to the now-populated index.
+        self._service.write_pointer_set(self._revision)
+        # Set the read pointer to point at the now-populated index
+        self._service.read_pointer_set(self._revision)
```

Comment:
Kind of inconsequential since the `refresh()` method affects indices and the `read_pointer_set` affects aliases. The index that had new documents added would get affected by the `refresh()` regardless of the alias, since the API targets the entire cluster.